README.md
# SyncMediaTrack
Synchronize Media Data from track GPX

Getting started with SyncMediaTrack

## 1) Set-up your Camera

Set the local time and date of your camera precisely before shooting.

Alternatively you can set your camera time to GMT (http://wwp.greenwichmeantime.com/) which can be handy as you won’t have to set summer/winter time or when you travel through time zones.

## 2) Take pictures while recording a tracklog

Make sure your GPS receiver is recording a track log. Keep your GPSr ON during all the time you take pictures.

## 3) Download exiftool 

Get the latest version from https://exiftool.org/ or install it from the installer of your Linux distribution.
In **Windows** you need to copy the _exiftool_ executable to some directory included in the %PATH% environment variable, for example c:\Windows.

## 4) Download ( ffmpeg / ffprobe ) (optional)

Ffmpeg is required to read the GPS latitude / longitude and GPSDatetime from the video GoPro files.
Get the latest version from https://ffmpeg.org/ or install it from the installer of your Linux distribution.
In **Windows** you need to copy the _ffmpeg_ and _ffprobe_ executable to some directory included in the %PATH% environment variable, for example c:\Windows.

## 5) Run from command-line SyncMediaTrack
First it is advisable to check that the images are well located,
```
SyncMediaTrack updatemedia --dry-run --geoservice --track XXXX.gpx photos/Andorra
```
Once we have verified that everything is correct we can execute it again adding the correct geographic positions
```
SyncMediaTrack updatemedia --track XXXX.gpx photos/Andorra
```

# Reorganize your tracks

If you have several tracks you can reorganize them chronologically and geolocalized with the following command
```
SyncMediaTrack updatetrack --track <trackdir or gpx file> ---geoservice 
```

### show results

```
[2023-03-26_09-59_Sun.gpx] -> 2023_03_26_09_59_mon_Fanlo Aragón España.gpx
[Les preses limpio.gpx] -> 2023_03_05_09_46_mon_la Vall d'en Bas España.gpx
[Move_2019_09_08_17_47_49_Ciclismo.gpx] -> 2019_09_08_17_48_mon_Orpesa _ Oropesa del Mar España.gpx
[Mulacen2.gpx] -> No time found in GPX file
[TrackWaypoint.gpx] -> No time found in GPX file
[benicasim-cueva-de-la-seda-23-06-2020.gpx] -> 2020_06_23_08_41_mon_Benicàssim _ Benicasim España.gpx
[cedraman-castillo-de-villamalefa-por-gp-rio-villahermosa-ced.gpx] -> 2019_06_29_10_11_mon_Castillo de Villamalefa España.gpx
```

---

# Trouble Shooting

#### No indications of localization appears

Check your gpx file first then your pictures this way.

You can open and modify your GPX file easily in a text editor: it has plenty of tags
```
<tag>
....
</tag>
```

Gpx files have one or more

```
<trkseg>
...
</trkseg>
```
which contain plenty of trackpoints
```
<trkpt>
...
</trkpt>
```
which says at which location you were at a precise time.
```
...
<trkseg>
  ... 
  <trkpt lat="48.50517319" lon="7.13916969">
    <ele>700.46</ele>
    <time>2007-03-04T12:05:16Z</time>
  </trkpt>
  <trkpt lat="48.50517286" lon="7.13916759">
    <ele>704.30</ele>
    <time>2007-03-04T12:05:18Z</time>
  </trkpt>
  <trkpt lat="48.50517294" lon="7.13916550">
    <ele>708.15</ele>
    <time>2007-03-04T12:05:20Z</time>
  </trkpt
  ... 
</trkseg>
...
```

# TODO

* [ ] Add support for read GPS Time from Gopro Video
* [ ] Fix GPX file when position is more than 500m away
* [ ] Fix GPX file when date mismatch or disordered
* [ ] Update elevation from DEM (https://github.com/lukeroth/gdal)
* [ ] Support viper for configuration files (https://github.com/spf13/viper)
* [ ] Sort tracks by type of activity, hiking, running, cycling


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

-

## [1.3] - 2023-05-04

### Fixed

- Show Warning if GPX file does not have time stamps
- Fix bad condition to get closes GPS
- Fix elevation when has decimal values
- Fix GPX file has time stamps out of order
- Fix GPX timezone

### Added

- Show a notice when gps data is updated
- Add an option to show more verbose information
- Add default country support to remove geolocation
- Update tracks with geolocation, date and time
- Add support for store the old filename in header.name
- Check if the gpx file has positions further away than 1km
- Adds support to search images with GPS positioning too
- Warns if the distance is greater than 500 meters in less than 30 seconds

## [1.2] - 2023-05-01

### Fixed

- Restructures the code in directories
- Fix use of global variable
- Fix problem when exif data is not obtained if GPS time is obtained from mp4

### Added

- Include support for read GPS Time from Gopro Video
- Support for read any video or image file
- show date comparison
- Get GPS Time from image file
- Show kind of time (Archive time, Exit time, GPS time)

## [1.1] - 2023-04-20

### Fixed

- Fix load GPX files
- Fix store elevation
- Update dependencies
- Fix small errors

### Added

- Support for geocoding using openstreetmap
- New parameter to force update even overwriting previous GPS data
- Support for multiples GPX files
- Use color red for fatal errors
- Show old location
- Support to adjust the date according to the time zone and if it has daylight saving time

## [1.0] - 2023-04-15

- Initial version

[Unreleased]: https://github.com/inode64/SyncMediaTrack/compare/v1.1...main
[1.1]: https://github.com/inode64/SyncMediaTrack/compare/v1.0...v1.1
[1.2]: https://github.com/inode64/SyncMediaTrack/compare/v1.1...v1.2
[1.3]: https://github.com/inode64/SyncMediaTrack/compare/v1.2...v1.3


SyncMediaTrack.go
package main

import (
	"encoding/xml"
	"errors"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"time"

	"github.com/barasher/go-exiftool"
	"github.com/codingsince1985/geo-golang/openstreetmap"
	"github.com/fatih/color"
	"github.com/gabriel-vasile/mimetype"
	"github.com/karrick/godirwalk"
	"github.com/spf13/cobra"
)

type Gpx struct {
	XMLName xml.Name `xml:"gpx"`
	Trk     Trk      `xml:"trk"`
}

type Trk struct {
	Trkseg Trkseg `xml:"trkseg"`
}

type Trkseg struct {
	Trkpt []Trkpt `xml:"trkpt"`
}

type Trkpt struct {
	Lat  float64 `xml:"lat,attr"`
	Lon  float64 `xml:"lon,attr"`
	Time string  `xml:"time"`
	Ele  float64 `xml:"ele"`
}

var (
	dataGPX    []Gpx
	dryRun     bool
	force      bool
	geoservice bool
	mediaDir   string
	track      string
)

	colorGreen  = color.New(color.FgGreen).SprintFunc()
	colorRed    = color.New(color.FgRed).SprintFunc()
	colorYellow = color.New(color.FgYellow).SprintFunc()
)

var rootCmd = &cobra.Command{
	Use:     "SyncMediaTack",
	Short:   "Synchronize Media Data from track GPX",
	Long:    `Using a gpx track, analyze a directory with images or movies and add the GPS positions`,
	Args:    cobra.MinimumNArgs(1),
	Version: "1.1",
	Run: func(cmd *cobra.Command, args []string) {
		if len(args) == 1 {
			mediaDir = args[0]
		}
	},
}

func init() {
	rootCmd.PersistentFlags().BoolVar(&dryRun, "dry-run", false, "Performs the actions without writing to the files")
	rootCmd.PersistentFlags().BoolVar(&force, "force", false, "Force update even overwriting previous GPS data")
	rootCmd.PersistentFlags().BoolVar(&geoservice, "geoservice", false, "Show location from GPS position from geocoding service of openstreetmap")
	rootCmd.PersistentFlags().StringVar(&track, "track", "", "GPX track or a directory of GPX tracks")
}

func ReverseLocation(location Trkpt) (string, error) {
	service := openstreetmap.Geocoder()

	address, err := service.ReverseGeocode(location.Lat, location.Lon)
	if err != nil {
		return "", err
	}

	if len(address.City) < 9 && address.State != "" {
		return fmt.Sprintf("%s %s %s", address.City, address.State, address.Country), nil
	}

	return fmt.Sprintf("%s %s", address.City, address.Country), nil
}

func readGPX(filename string) {
	fmt.Printf("Processing: %v \n", filename)

	mtype, err := mimetype.DetectFile(filename)
	if err != nil {
		log.Fatal(colorRed(err))
	}

	if !mtype.Is("application/gpx+xml") && !mtype.Is("text/xml") {
		return
	}

	file, err := os.Open(filename)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	defer file.Close()

	gpx := Gpx{}
	decoder := xml.NewDecoder(file)
	if err := decoder.Decode(&gpx); err != nil {
		fmt.Println(colorYellow("Warning: GPX file could not be processed, error: ", colorRed(err)))
		return
	}

	dataGPX = append(dataGPX, gpx)
}

func readGPXDir(trackDir string) {
	err := godirwalk.Walk(trackDir, &godirwalk.Options{
		Callback: func(path string, de *godirwalk.Dirent) error {
			if de.IsDir() {
				return nil // do not remove directory that was provided top-level directory
			}

			readGPX(path)

			return nil
		},
		Unsorted: true,
	})
	if err != nil {
		fmt.Println(err)
	}
}

func GetMediaDate(filename string, t *time.Time, gps *Trkpt) error {
	f, err := os.Stat(filename)
	if err != nil {
		return err
	}
	*t = f.ModTime()

	// create an instance of exiftool
	et, err := exiftool.NewExiftool(exiftool.CoordFormant("%+f"))
	if err != nil {
		return err
	}
	defer et.Close()

	metas := et.ExtractMetadata(filename)

	gps.Lon, _ = metas[0].GetFloat("GPSLongitude")
	gps.Lat, _ = metas[0].GetFloat("GPSLatitude")
	EleStr, err := metas[0].GetString("GPSAltitude")
	if err == nil {
		re := regexp.MustCompile(`(\d+) m.*`)
		match := re.FindStringSubmatch(EleStr)

		if len(match) > 1 {
			alt, err := strconv.Atoi(match[1])
			if err == nil {
				gps.Ele = float64(alt)
			}
		}
	}

	// define the list of possible tags to extract date from
	dateTags := []string{"DateTimeOriginal", "DateTime", "DateTimeDigitized"}

	// loop through the tags until a valid date is found
	for _, tag := range dateTags {
		val, err := metas[0].GetString(tag)
		if err != nil {
			continue
		}
		if val != "" {
			date, err := time.Parse("2006:01:02 15:04:05", val)
			if err != nil {
				continue
			}
			*t = date
			return nil
		}
	}

	return nil
}

func GetClosesGPS(imageTime time.Time) (Trkpt, error) {
	var closestPoint Trkpt
	var closestDuration time.Duration

	for _, gpx := range dataGPX {
		for _, trkpt := range gpx.Trk.Trkseg.Trkpt {
			if len(trkpt.Time) == 0 {
				continue
			}
			trkptTime, err := time.Parse("2006-01-02T15:04:05Z", trkpt.Time)
			if err != nil {
				fmt.Printf(colorRed(err) + "\n")
				continue
			}

			duration := imageTime.Sub(trkptTime.UTC())
			if duration < 0 {
				duration = -duration
			}

			if closestDuration == 0 || duration < closestDuration {
				closestPoint = trkpt
				closestDuration = duration
			}
		}

		if closestDuration.Seconds() > 30 {
			return closestPoint, errors.New("There is no close time to obtain the GPS position")
		}
	}

	return closestPoint, nil
}

func WriteGPS(gps Trkpt, filename string) error {
	et, err := exiftool.NewExiftool()
	if err != nil {
		return err
	}
	defer et.Close()

	// Extract file metadata
	fileInfos := et.ExtractMetadata(filename)
	if len(fileInfos) == 0 {
		return fmt.Errorf("no metadata found %s", filename)
	}
	fileInfo := fileInfos[0]
	if fileInfo.Err != nil {
		return fileInfo.Err
	}

	latRef := "N"
	if gps.Lat >= 0 {
		latRef = "S"
	}
	lonRef := "E"
	if gps.Lon >= 0 {
		lonRef = "W"
	}

	gpsTime, err := time.Parse("2006-01-02T15:04:05Z", gps.Time)
	if err != nil {
		log.Fatal(colorRed(err))
	}

	fileInfo.SetString("GPSDateStamp", gpsTime.Format("2006:01:02"))
	fileInfo.SetString("GPSTimeStamp", gpsTime.Format("15:04:05,00"))

	// Update latitude, longitude, and elevation values
	fileInfo.SetFloat("GPSLatitude", gps.Lat)
	fileInfo.SetFloat("GPSLongitude", gps.Lon)
	fileInfo.SetInt("GPSAltitude", int64(gps.Ele))
	fileInfo.SetString("GPSAltitudeRef", "above sea level")

	fileInfo.SetString("GPSLatitudeRef", latRef)
	fileInfo.SetString("GPSLongitudeRef", lonRef)

	// Write the new metadata to the file.
	et.WriteMetadata([]exiftool.FileMetadata{fileInfo})

	return nil
}

func fileIsMedia(filename string) bool {
	allowed := []string{"image/jpeg", "video/mp4"}

	mtype, err := mimetype.DetectFile(filename)
	if err != nil {
		return false
	}

	return mimetype.EqualsAny(mtype.String(), allowed...)
}

func main() {
	var gpsOld Trkpt

	cobra.CheckErr(rootCmd.Execute())

	fileInfo, err := os.Stat(track)
	if err != nil {
		log.Fatal(colorRed("No open GPX path"))
	}

	if fileInfo.IsDir() {
		readGPXDir(track)
	} else {
		readGPX(track)
	}

	if len(dataGPX) == 0 {
		log.Fatal(colorRed("There is no track processed"))
	}

	err = godirwalk.Walk(mediaDir, &godirwalk.Options{
		Callback: func(path string, de *godirwalk.Dirent) error {
			var date time.Time

			if de.IsDir() {
				return nil // do not remove directory that was provided top-level directory
			}

			if !fileIsMedia(path) {
				return nil
			}

			relPath, err := filepath.Rel(mediaDir, path)
			if err != nil {
				return err
			}
			fmt.Printf("[%v] - ", relPath)

			err = GetMediaDate(path, &date, &gpsOld)
			if err != nil {
				fmt.Println(err)
				return nil
			}

			location, err := GetClosesGPS(date)
			if err != nil {
				fmt.Println(colorRed(err))
				return nil
			}

			if gpsOld.Lat == 0 && gpsOld.Lon == 0 {
				fmt.Printf("No location")
			} else {
				fmt.Printf("Lat %v Lon %v Ele %v", gpsOld.Lat, gpsOld.Lon, gpsOld.Ele)
			}

			fmt.Printf(" -> Lat %v Lon %v Ele %v ", location.Lat, location.Lon, location.Ele)

			if geoservice {
				loc, _ := ReverseLocation(location)
				if len(loc) != 0 {
					fmt.Printf("(%s)", colorGreen(loc))
				}
			}
			if !force && gpsOld.Lat != 0 && gpsOld.Lon != 0 {
				color.Yellow("(no update)")
				return nil
			}

			fmt.Printf("\n")

			if dryRun {
				return nil
			}

			err = WriteGPS(location, path)
			if err != nil {
				fmt.Println(err)
				return nil
			}

			return nil
		},
	})
	if err != nil {
		fmt.Println(err)
	}
}
